[RFC] directory structure for architecture-specific code and headers

bkueng opened the pull request:
Currently our architecture-specific code is not in a common place, and not handled consistently either. And the upcoming stm32h7 will come with more differences, so this is a good reason to get to a better structure.

This PR introduces the following directory structure under `platforms`:
```
platforms/nuttx/src/px4> tree
.
├── CMakeLists.txt
├── common
│   ├── CMakeLists.txt
│   ├── console_buffer.cpp
│   ├── px4_init.cpp
│   ├── px4_nuttx_impl.cpp
│   ├── px4_nuttx_tasks.c
│   └── px4_platform_micro_hal.h
├── nxp
│   └── kinetis
│       └── px4_arch_micro_hal.h
└── stm
    ├── CMakeLists.txt
    ├── stm32_common
    │   ├── px4_arch_micro_hal.h
    │   ├── px4io_serial
    │   │   └── arch_px4io_serial.h
    │   └── tone_alarm
    │       ├── CMakeLists.txt
    │       ├── ToneAlarmInterface.cpp
    │       ├── ToneAlarmInterfaceGPIO.cpp
    │       └── ToneAlarmInterfacePWM.cpp
    ├── stm32f1
    │   ├── CMakeLists.txt
    │   └── px4_arch_micro_hal.h
    ├── stm32f4
    │   ├── CMakeLists.txt
    │   ├── px4_arch_micro_hal.h
    │   └── px4io_serial
    │       ├── arch_px4io_serial.h
    │       ├── CMakeLists.txt
    │       └── px4io_serial.cpp
    └── stm32f7
        ├── CMakeLists.txt
        ├── px4_arch_micro_hal.h
        ├── px4io_serial
        │   ├── arch_px4io_serial.h
        │   ├── CMakeLists.txt
        │   └── px4io_serial.cpp
        └── tone_alarm
            ├── CMakeLists.txt
            ├── ToneAlarmInterface.cpp
            ├── ToneAlarmInterfaceGPIO.cpp
            └── ToneAlarmInterfacePWM.cpp
```
Goals:
- reduce amount of per-board configuration and use cmake deps to automatically pull in what is needed.
- provide a generic & flexible structure
- allow chip-specific custom code/headers
- allow code sharing between similar chips
- reduce amount of ifdefs

The main structure takes the form `<chip_manufacturer>/<chip>`. Both of these need to be specified in the board config. The chip-specific directory then defines whether it includes common code or provides its own implementation. For example for `tone_alarm`, it provides an architecture-specific library, which the tone_alarm module depends on, so that it's automatically linked if the board adds `tone_alarm`.

State of the PR:
- (partially) done: tone_alarm, px4io_serial, px4_micro_hal
- It builds on SITL, v2, v4 and v5

dagar commented on boards/px4/fmu-v3/default.cmake:
```diff
@@ -8,6 +8,8 @@ px4_add_board(
 	TOOLCHAIN arm-none-eabi
 	ARCHITECTURE cortex-m4
+	CHIP_MANUFACTURER stm
+	CHIP stm32f4
```
We should be able to get this from the NuttX config.
 - CONFIG_ARCH_CHIP
 - CONFIG_ARCH_CHIP_STM32
 - CONFIG_STM32_STM32F4XXX

dagar replied:
https://github.com/PX4/Firmware/blob/f3eafdc296a8b0413283916ad80450b668b027e4/src/drivers/boards/common/CMakeLists.txt#L44-L48

davids5 replied:
why not add_subdirectory(${CONFIG_ARCH_CHIP})

bkueng replied:
The issue with this is that `CONFIG_ARCH_CHIP` gives `stm32` and not `stm32f4`. I did not see anything that would provide the stm32f4, but maybe I missed it?

davids5 replied:
> but maybe I missed it?
@bkueng - welcome to the world of compressed configs.  Easy to compare but suffer from information hiding. 

You have to look in the .config or config.h file there will be something in the form CONFIG_STM32_STM32F4XXX=y  the we will have to translate and concatenate

bkueng replied:
As discussed with David, personally I generally like to infer this as well, but we both think it's better here not to, because:
- It's not generic for other platforms
- We would need to have NuttX-config specific if-else cases to translate to our directories and extend it every time for a new arch/chip.
- Stating the arch/chip explicitly makes it easier to see what code is used.

@dagar what do you think?

bkueng replied:
Alright, changed.

davids5 commented on platforms/nuttx/src/px4/nxp/kinetis/px4_arch_micro_hal.h:
```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+
+#include "../../common/px4_platform_micro_hal.h"
```
Can we invert this or am I thinking about it wrong?

px4_platform_micro_hal.h 
includes px4_arch_micro_hal.h and cmake sets the path to <mfg>/<chip>

as an alternate the #if defined() pattern used in nuttx work be ok - albeit a point of collision on merges. It is just clear

bkueng replied:
What is your reason for inverting? I'm fine with either way, this just gives a bit more flexibilty: the arch-specific header could define something that is used in the platform-specific include (not needed now, but maybe in future).

davids5 replied:
I am not sure I understand the ramification well enough, maybe we can discuss it. I was looking at it from what it was (common interface) provide to PX4 that the specifics were conditioned on ARCH/CHIP type.

davids5 commented on platforms/nuttx/src/px4/stm/stm32_common/px4_arch_micro_hal.h:
```diff
@@ -0,0 +1,105 @@
+ * used for selection of significant digits of the UUID in the PX4 code base.
+ * This is done to avoid the ripple effects changing the IDs used on STM32 base platforms
+ */
+#if defined(PX4_CPU_UUID_CORRECT_CORRELATION)
```
Just a side note:we need to eradicate this. - It is a hack for external API that used to narrow values for uniqueness. The issue is the not all vendors will document the sequence

bkueng commented:
I restructured the include paths to make them more explicit and avoid the 'px4_arch_' filename prefix:
```
.
├── CMakeLists.txt
├── stm32_common
│   ├── adc
│   │   ├── adc.cpp
│   │   └── CMakeLists.txt
│   ├── hrt
│   │   ├── CMakeLists.txt
│   │   └── hrt.c
│   ├── include
│   │   └── px4_arch
│   │       ├── adc.h
│   │       ├── io_timer.h
│   │       ├── micro_hal.h
│   │       └── px4io_serial.h
│   ├── io_pins
│   │   ├── CMakeLists.txt
│   │   ├── input_capture.c
│   │   ├── io_timer.c
│   │   ├── pwm_servo.c
│   │   └── pwm_trigger.c
│   ├── led_pwm
│   │   ├── CMakeLists.txt
│   │   └── led_pwm.cpp
│   └── tone_alarm
│       ├── CMakeLists.txt
│       ├── ToneAlarmInterface.cpp
│       ├── ToneAlarmInterfaceGPIO.cpp
│       └── ToneAlarmInterfacePWM.cpp
├── stm32f1
│   ├── CMakeLists.txt
│   └── include
│       └── px4_arch
│           └── micro_hal.h
├── stm32f4
│   ├── CMakeLists.txt
│   ├── include
│   │   └── px4_arch
│   │       ├── adc.h
│   │       ├── io_timer.h
│   │       ├── micro_hal.h
│   │       └── px4io_serial.h
│   └── px4io_serial
│       ├── CMakeLists.txt
│       └── px4io_serial.cpp
└── stm32f7
    ├── CMakeLists.txt
    ├── include
    │   └── px4_arch
    │       ├── adc.h
    │       ├── io_timer.h
    │       ├── micro_hal.h
    │       └── px4io_serial.h
    └── px4io_serial
        ├── CMakeLists.txt
        └── px4io_serial.cpp

19 directories, 37 files
```

@davids5 I'll take a look at the proper kinetis dir names.

bkueng commented:
Updates:
- moved everything from `src/platforms` to `platforms` under the right directories. This results in quite a large diff due to the include changes (https://github.com/PX4/Firmware/pull/12797/commits/995c4141ce745f25e19e1ba3c1cb4b39a864676c). It's IMO cleaner but we can discuss if we want that. It also requires submodule changes (ecl, driver framework, matrix).
- similar for `src/drivers/drv_hrt.h`: moved to platforms, resulting in an include change (https://github.com/PX4/Firmware/pull/12797/commits/abffa7ebe2350bf5b28f29b6f5d6c3d3bf346917). I'm not quite sure if we should move the rest of `src/drivers/drv_*` to `platforms/common/include/px4_platform_common/drivers`, since some of them also define platform-specific API's.

Next steps:
- update all board configs
- submodule updates

What's still left afterwards is `src/drivers/boards/common`.

bkueng commented:
Sure, on the chip-level nothing changed anymore, just on the platform level:
```
platforms/common> tree
.
├── apps.cpp.in
├── apps.h.in
├── CMakeLists.txt
├── empty.c
├── include
│   └── px4_platform_common
│       ├── app.h
│       ├── atomic.h
│       ├── cli.h
│       ├── config.h
│       ├── console_buffer.h
│       ├── defines.h
│       ├── getopt.h
│       ├── hrt.h
│       ├── i2c.h
│       ├── init.h
│       ├── log.h
│       ├── micro_hal.h
│       ├── module.h
│       ├── module_params.h
│       ├── param.h
│       ├── param_macros.h
│       ├── posix.h
│       ├── px4_work_queue
│       │   ├── ScheduledWorkItem.hpp
│       │   ├── WorkItem.hpp
│       │   ├── WorkQueue.hpp
│       │   └── WorkQueueManager.hpp
│       ├── sem.h
│       ├── sem.hpp
│       ├── shmem.h
│       ├── shutdown.h
│       ├── tasks.h
│       ├── time.h
│       └── workqueue.h
├── module.cpp
├── px4_cli.cpp
├── px4_getopt.c
├── px4_log.c
├── px4_work_queue
│   ├── CMakeLists.txt
│   ├── ScheduledWorkItem.cpp
│   ├── test
│   │   ├── CMakeLists.txt
│   │   ├── wqueue_main.cpp
│   │   ├── wqueue_scheduled_test.cpp
│   │   ├── wqueue_scheduled_test.h
│   │   ├── wqueue_start.cpp
│   │   ├── wqueue_test.cpp
│   │   └── wqueue_test.h
│   ├── WorkItem.cpp
│   ├── WorkQueue.cpp
│   └── WorkQueueManager.cpp
├── shutdown.cpp
└── work_queue
    ├── CMakeLists.txt
    ├── dq_addlast.c
    ├── dq_rem.c
    ├── dq_remfirst.c
    ├── hrt_queue.c
    ├── hrt_thread.c
    ├── hrt_work_cancel.c
    ├── queue.c
    ├── sq_addafter.c
    ├── sq_addlast.c
    ├── sq_remfirst.c
    ├── work_cancel.c
    ├── work_lock.c
    ├── work_lock.h
    ├── work_queue.c
    ├── work_thread.c
    └── wqueue_test
        ├── CMakeLists.txt
        ├── wqueue_main.cpp
        ├── wqueue_start_posix.cpp
        ├── wqueue_test.cpp
        └── wqueue_test.h

```

hamishwillee commented:
@bkueng When this delivers, can you update http://dev.px4.io/master/en/hardware/porting_guide.html#flight-controller-configuration-file-layout (and anywhere else you think is relevat)?

bkueng commented:
> @bkueng When this delivers, can you update http://dev.px4.io/master/en/hardware/porting_guide.html#flight-controller-configuration-file-layout (and anywhere else you think is relevant)?

I don't see anything there that would be affected by this change.
